vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -81,46 +98,163 @@ def create_weights(
+                    act_scale, {
+                        **extra_weight_attrs,
+                        "shard_indexer":
+                        self.scales_shard_indexer,
+                    })
+
+    def scales_shard_indexer(
+            self, param: torch.Tensor, loaded_weight: torch.Tensor,
+            shard_id: Union[str, int]) -> Tuple[torch.Tensor, torch.Tensor]:
+        qkv_idxs = {"q": 0, "k": 1, "v": 2}
+
+        if isinstance(shard_id, int):
+            pass
+        elif isinstance(shard_id, str):
+            if shard_id not in qkv_idxs:
+                raise ValueError(f"Unknown shard_id: {shard_id}")
+            shard_id = qkv_idxs[shard_id]
+        else:
+            ValueError(f"Shard id must be int or str but got {type(shard_id)}")
+
+        return param[shard_id], loaded_weight
 
     def process_weights_after_loading(self, layer: Module) -> None:
-        # Although the quant_method is propagated to all layers,
-        # only linear layers invoke "create_weights". So we check
-        # whether "weight_scaling_facor" is registered to determine
-        # whether the layer is a linear layer that requires quantization.
-        if not hasattr(layer, "weight_scaling_factor"):
+        if (not hasattr(layer, "process_after_load")
+                or not layer.process_after_load):
             return
 
-        qweight, weight_scale = ops.scaled_fp8_quant(layer.weight)
-        # torch._scaled_mm requires column-major in the second
-        # input (weight), so we transpose the quantized weight.
-        layer.weight = Parameter(qweight.t(), requires_grad=False)
-        layer.weight_scaling_factor.data.copy_(weight_scale)
+        # If checkpoint is fp/bf16 (not serialized fp8), quantize the weights.
+        if not self.quant_config.is_serialized:
+            qweight, weight_scale = ops.scaled_fp8_quant(layer.weight,
+                                                         scale=None)
+            layer.weight = Parameter(qweight.t(), requires_grad=False)
+            layer.weight_scale = Parameter(weight_scale, requires_grad=False)
+            layer.logical_widths = None
+            layer.act_scale = None
+            return
+
+        # TODO: cutlass kernels will remove the need for much of this logic.
+        # If the checkpoint is serialized fp8, we already loaded quantized,
+        # so, just cleanup the Parameters for easier use in apply().
+        else:
+            # WEIGHT
+            #   Transpose weight for passing to torch._scaled_mm
+            weight = layer.weight
+            layer.weight = Parameter(weight.t(), requires_grad=False)
+
+            # WEIGHT_SCALE
+            #   If all weight_scales are equal, use a single scale.
+            if all_close_1d(layer.weight_scale):
+                layer.weight_scale = Parameter(layer.weight_scale.max(),
+                                               requires_grad=False)
+                layer.logical_widths = None
```

Comment:
Can you elaborate this a bit? IIUC, this is used when the weights were quantized per-channel, but since their the scaling factors are similar so we can actually use per-tensor during execution?

Reply:
Right now, we have a naive for loop to handle N scales for QKV / GateUpProj.

I made a few checkpoints where I quantized `Wq`, `Wk`, and `Wk` as well as `Wgate` and `Wup` jointly (such that we can do one call to `torch._scaled_mm` without needing to loop

So this step auto-detects if the checkpoint has shared scales and uses them.

Once we have the cutlass kernels working, we can likely remove this.

Reply:
Make sense. Thanks for clarification! If possible, we could add a comment for future reference.